src/stransport_stream.c

```diff
@@ -150,9 +156,9 @@ static OSStatus read_cb(SSLConnectionRef conn, void *data, size_t *len)
 	requested = left = *len;
 	do {
 		ret = git_stream_read(io, data + (requested - left), left);
-		if (ret < 0) {
+		if (ret <= 0) { /* There was an error or nothing could be read because stream is closed */
 			*len = 0;
-			return -1;
+			return errSSLClosedAbort;
```

Comment:
Finally you can see how VLC does it here: https://github.com/videolan/vlc/blob/8a7b1417015346444b6d0a329a95a1dada535429/modules/misc/securetransport.c.

It does return `errSSLClosedGraceful` when reading 0, so a different behavior, and I would assume their code has been tested and deployed extensively.

This whole thing is pretty messy...